pkg/local_object_storage/engine/select.go

```diff
@@ -49,3 +52,94 @@ func (e *StorageEngine) Select(cnr cid.ID, filters object.SearchFilters) ([]oid.
 
 	return addrList, nil
 }
+
+// List returns `limit` available physically storage object addresses in engine.
+// If limit is zero, then returns all available object addresses.
+//
+// Returns an error if executions are blocked (see BlockExecution).
+func (e *StorageEngine) List(limit uint64) ([]oid.Address, error) {
+	if e.metrics != nil {
+		defer elapsed(e.metrics.AddListObjectsDuration)()
+	}
+
+	e.blockMtx.RLock()
+	defer e.blockMtx.RUnlock()
+
+	if e.blockErr != nil {
+		return nil, e.blockErr
+	}
+
+	addrList := make([]oid.Address, 0, limit)
+	uniqueMap := make(map[string]struct{})
+	ln := uint64(0)
+
+	// consider iterating over shuffled shards
+	for _, sh := range e.unsortedShards() {
+		res, err := sh.List() // consider limit result of shard iterator
+		if err != nil {
+			e.reportShardError(sh, "could not select objects from shard", err)
+			continue
+		}
+		for _, addr := range res { // save only unique values
+			if _, ok := uniqueMap[addr.EncodeToString()]; !ok {
+				uniqueMap[addr.EncodeToString()] = struct{}{}
+				addrList = append(addrList, addr)
+
+				ln++
+				if limit > 0 && ln >= limit {
+					return addrList, nil
+				}
+			}
+		}
+	}
+
+	return addrList, nil
+}
+
+// Search performs Search op on all underlying shards and returns merged result.
+//
+// Fails instantly if executions are blocked (see [StorageEngine.BlockExecution]).
+func (e *StorageEngine) Search(cnr cid.ID, fs object.SearchFilters, attrs []string, cursor *meta.SearchCursor, count uint16) ([]client.SearchResultItem, *meta.SearchCursor, error) {
+	if e.metrics != nil {
+		defer elapsed(e.metrics.AddSearchDuration)() // TODO: sep metric?
+	}
+	e.blockMtx.RLock()
+	defer e.blockMtx.RUnlock()
+	if e.blockErr != nil {
+		return nil, nil, e.blockErr
+	}
+	shs := e.unsortedShards()
+	if len(shs) == 0 {
+		return nil, nil, nil
+	}
+	items, nextCursor, err := shs[0].Search(cnr, fs, attrs, cursor, count)
+	if err != nil {
+		e.reportShardError(shs[0], "could not select objects from shard", err)
+	}
+	if len(shs) == 1 {
+		return items, nextCursor, nil
+	}
+	shs = shs[1:]
+	mrg := objectcore.NewSearchResultMerger(count)
+	mrg.Add(nextCursor != nil, items)
+	for i := range shs {
+		// TODO: avoid allocation inside and pass pre-allocated buffer
```

Comment:
We can afford 1K elements buffer. Usually it's some KBs for memory. Likely you won't have 100 shards. It can be optimized a little with iterators where you'd picking elements one by one from respective shards, but let's keep it simple for now.

Reply:
droppped